### PR TITLE
RSDK-8087: add wrappers over registry methods + resolve cyclic dependencies 

### PR DESCRIPTION
**Pull request by puranamp**
This change creates exported wrapper functions on the logger manager instance methods. This allows us to utilize these instance methods outside the logging package. After talking offline, it was decided that keeping the logger manager private but having publicly exposed functions that operate on the manager is the best approach. As a side note, I left the instance methods as is since they are used for testing.

**UPDATE**:
Now closes [this](https://viam.atlassian.net/browse/RSDK-8087) ticket. 

### Changes

- Moved `log_config.go` and corresponding tests into the logging package. This reduces the likelihood of cyclic dependencies arising between the `logging` and `config` packages and localized the `UpdateRegistry` function near the registry code.
- `LoggerPatternConfig` now lives inside the `logging` package. This is necessary to prevent cyclic dependencies.
- Call UpdateLoggerRegistry inside `reader.go`
- new test in `reader_test.go` ensuring that logger config is read and utilized correctly.
- `DeRegisterLogger` and `GetRegisteredLoggerNames` methods in the logger manager

### CODE REVIEW

**Comment by benjirewis** [comment]
> After talking offline, it was decided that keeping the logger manager private but having publicly exposed functions that operate on the manager is the best approach. As a side note, I left the instance methods as is since they are used for testing.

Why not just make `LoggerRegistry` public, keep its fields private, and export the getters and setters you need?

**Comment by puranamp** [comment]
> Why not just make `LoggerRegistry` public, keep its fields private, and export the getters and setters you need?

We want to limit the scope of the LoggerRegistry variable to be just inside the RDK. I tried making an internal package but that didn't work out so I went with this approach.

**Comment by puranamp** [comment]
> What other packages need to access the logger manager?

two packages, `logging` and `config`

File: logging/logger_manager.go [hunk 1]
```diff
@@ -41,3 +41,20 @@ func (lr *loggerRegistry) updateLoggerLevel(name string, level Level) error {
 	logger.SetLevel(level)
 	return nil
 }
+
+// Exported Functions specifically for use on global logger manager.
```
Comment (benjirewis) [inline logging/logger_manager.go:45]: Could we add a `TODO` and a ticket number above the declaration of `loggerManager` to make the `loggerManager` not a global and instead a variable on `localRobot`? Or, make it a nonglobal as part of this PR if it's a somewhat trivial change?

**Comment by benjirewis** [comment]
>We want to limit the scope of the LoggerRegistry variable to be just inside the RDK. I tried making an internal package but that didn't work out so I went with this approach.

Discussed offline: this PR is about making access to the `loggerManager` public. The `loggerRegistry` type and it visibility are sort of irrelevant for the time being. My bad.

File: logging/logger_manager.go (hunk 1 above)
Comment (puranamp) [reply logging/logger_manager.go:45]: the `loggerManager` is already non-global, we just have functions that operate specifically on that variable. I'll make a ticket about making it a variable on `localRobot`

Comment (puranamp) [reply logging/logger_manager.go:46]: done!

**Review by benjirewis — APPROVED** [review]

**Comment by maximpertsov** [comment]
> > What other packages need to access the logger manager?
> 
> two packages, `logging` and `config`. maybe even `resource`?

Would it be possible to make some of those changes as part of this PR? I think it would help validate how we've decided to expose the log registry manager.

**Review by maximpertsov — CHANGES_REQUESTED** [review]
RE: https://github.com/viamrobotics/rdk/pull/4204#issuecomment-2233636607

I think it would be worth implementing some of the changes that depend on exposing these methods before we merge this change as is. Let me know if folks disagree here.

File: config/reader.go [hunk 2]
```diff
@@ -610,6 +610,10 @@ func processConfig(unprocessedConfig *Config, fromCloud bool, logger logging.Log
 		}
 	}
 
+	if err := logging.UpdateLoggerRegistry(cfg.LogConfig); err != nil {
+		return nil, err
+	}
```
Comment (maximpertsov) [inline config/reader.go:615]: let's add a small test for this in `config/reader_test.go` - e.g. reading a config with a `LogConfig` and ensuring that `GetRegisteredLoggerNames` returns the corresponding names if successful.

**Comment by danielbotros** [comment]
> RE: [#4204 (comment)](https://github.com/viamrobotics/rdk/pull/4204#issuecomment-2233636607)
> 
> I think it would be worth implementing some of the changes that depend on exposing these methods before we merge this change as is. Let me know if folks disagree here.

I think this is a good idea, or, at least trivially verifying that the changes made in this PR allow us to do what we want to do in `resource` and `config` so we don't have to muck with the registry too much more.

File: config/reader.go (hunk 2 above)
Comment (puranamp) [reply config/reader.go:615]: done!

**Comment by puranamp** [comment]
> I think this is a good idea, or, at least trivially verifying that the changes made in this PR allow us to do what we want to do in `resource` and `config` so we don't have to muck with the registry too much more.

As for configurations on resources, that is all done within the `config` package. I don't think we need to deal with the `resource` package as per my understanding after reading through the code again.

File: logging/logger_manager.go [hunk 3]
```diff
@@ -41,3 +52,40 @@ func (lr *loggerRegistry) updateLoggerLevel(name string, level Level) error {
...
+// RegisterLogger registers a new logger with a given name.
+func RegisterLogger(name string, logger Logger) {
+	loggerManager.registerLogger(name, logger)
+}
+
+// DeRegisterLogger attempts to remove a logger from the registry and returns a boolean denoting whether it succeeded.
+func DeRegisterLogger(name string) bool {
+	return loggerManager.deregisterLogger(name)
+}
+
```
Comment (maximpertsov) [inline logging/logger_manager.go:77]: per [this comment](https://github.com/viamrobotics/rdk/pull/4204#issuecomment-2236675867), do we still need these?

```suggestion
```

Comment (puranamp) [reply logging/logger_manager.go:77]: `RegisterLogger` is necessary, `DeRegisterLogger` is helpful for tests. I don't think there is a way to include a test in `reader_test.go` without this `DeRegisterLogger`.

Comment (maximpertsov) [reply logging/logger_manager.go:77]: ah sorry, missed that usage!

**Review by maximpertsov — APPROVED** [review]

**Review by danielbotros — APPROVED** [review]

File: logging/logger_manager.go [hunk 4]
```diff
@@ -41,3 +52,40 @@ func (lr *loggerRegistry) updateLoggerLevel(name string, level Level) error {
...
+// Exported Functions specifically for use on global logger manager.
+
+// RegisterLogger registers a new logger with a given name.
+func RegisterLogger(name string, logger Logger) {
+	loggerManager.registerLogger(name, logger)
+}
+
+// DeRegisterLogger attempts to remove a logger from the registry and returns a boolean denoting whether it succeeded.
+func DeRegisterLogger(name string) bool {
```
Comment (benjirewis) [inline logging/logger_manager.go:74]:
```suggestion
// DeregisterLogger attempts to remove a logger from the registry and returns a boolean denoting whether it succeeded.
func DeregisterLogger(name string) bool {
```
[super-nit] this is really one word.

**Review by benjirewis — APPROVED** [review]
LGTM just a spelling request.

**Pull request merged**